### PR TITLE
ragg2_fix

### PR DESCRIPTION
**Pull request by ZhangZhuoSJTU**
# egg_lang.c

  All the example tests are pull request to the radare2-regressions, in __radare2-regressions/bins/other/ragg2/__

## Bug Fixed:

1.  When get a local string variable includes special char like '\n', ragg2 will push a wrong string, because of forgetting the terminal NULL (example: bug1)
2.  When a string works as an argument of a function call, if the string includes ',', rcc pusharg will go wrong, because of ragg2 uses __strchr(ptr, ',')__ to separate arguments. (example: bug2)
3.  When a string starts with a non-zero number, the compiler will misunderstand it as a number, for the compiler uses __atoi(ptr)__ to determine it as a number. (example: bug3)
4.  "while" loop goes wrong because of the incorrect label. (example: bug4)
5.  "while" loop will cause following "if" frame wrong. (example: bug5)
6.  "while" and "if" goes wrong in such situation, the space before '{' is necessary, which is not a good syntax. (example: bug6)
```c
        wrong:
            if (4 < 5){ test(); }
        right:
            if (4 < 5) { test(); }
```

7.  Segmentfault when calling function with arguments in SYSCALL mode. (example: bug7)

## Support More:

1.  Support 'include' to use directory(example: include.r)
2.  Support 'if-else', which is necessary and important in a programming language, in one-pass compiler ragg2(example: ifelse.r)
3.  Support math operation in local variables assignment, which is also necessary. (example: mathop.r)

## Todo

1.  'div' and 'mul' instruction used in the wrong way in x86 and x64
2.  load_ptr in rcc_mathop goes wrong in arm

### CODE REVIEW

**Comment by ZhangZhuoSJTU** [comment]
Oh, I just edited file, libr/egg/egg_lang.c. Why is the check building always broken. So sad.

**Comment by radare** [comment]
doesnt builds

```
p/asm_dalvik.c: In function â€˜dalvik_disassembleâ€™:
p/asm_dalvik.c:13:35: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
p/../arch/hexagon/gnu/hexagon-dis.c: In function â€˜hexagon_dis_instâ€™:
p/../arch/hexagon/gnu/hexagon-dis.c:51:11: warning: format not a string literal and no format arguments [-Wformat-security]
p/../arch/hexagon/gnu/hexagon-opc.c: In function â€˜hexagon_get_opcode_machâ€™:
p/../arch/hexagon/gnu/hexagon-opc.c:1400:14: warning: unused variable â€˜mach_type_mapâ€™ [-Wunused-variable]
p/cs_mnemonics.c:1:14: warning: â€˜mnemonicsâ€™ defined but not used [-Wunused-function]
egg_lang.c: In function â€˜find_aliasâ€™:
egg_lang.c:253:5: error: â€˜forâ€™ loop initial declarations are only allowed in C99 mode
egg_lang.c:253:5: note: use option -std=c99 or -std=gnu99 to compile your code
egg_lang.c: In function â€˜r_egg_lang_parsecharâ€™:
egg_lang.c:1291:17: error: â€˜forâ€™ loop initial declarations are only allowed in C99 mode
egg_lang.c:1292:21: error: â€˜forâ€™ loop initial declarations are only allowed in C99 mode
make[3]: *** [egg_lang.o] Error 1
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [egg] Error 2
make[1]: *** [all] Error 2
make: *** [all] Error 2
```


please squash your commits, never merge in a PR

File: libr/egg/egg_lang.c [hunk 1]
```diff
@@ -3,67 +3,102 @@
 #include <r_egg.h>
 
 #define CTX egg->context
-char *nested[32] = {0};
-char *nestede[32] = {0};
+char *nested[32] = {NULL};
```
Comment (radare) [inline libr/egg/egg_lang.c]: please indent the file with r2-indent or using tabs with default indentation of vim, , the whole tab->space diff makes it impossible to review

File: libr/egg/egg_lang.c [hunk 2]
```diff
@@ -559,501 +767,569 @@ eprintf ("CNTXXXPUSH (%s)\n", ctxpush[context-1]);
...
+            if (elem_n) {
+                ptr = elem;
+                elem[elem_n] = '\0';
+                while (is_space (*ptr)) ptr++;
+                rcc_fun (egg, ptr);
+            }
+            elem_n = 0;
+        }
+    }
+    if (c!='\t' && c!=' ')
```
Comment (radare) [inline libr/egg/egg_lang.c]: missing braces {}

File: libr/egg/egg_lang.c [hunk 3]
```diff
@@ -559,501 +767,569 @@ eprintf ("CNTXXXPUSH (%s)\n", ctxpush[context-1]);
...
+            else elem[elem_n++] = c;
+            break;
+        default:
+            elem[elem_n++] = c;
+        }
+        if (slurp) {
+            if (elem_n) {
+                ptr = elem;
+                elem[elem_n] = '\0';
+                while (is_space (*ptr)) ptr++;
```
Comment (radare) [inline libr/egg/egg_lang.c]: missing braces

File: libr/egg/egg_lang.c [hunk 4]
```diff
@@ -559,501 +767,569 @@ eprintf ("CNTXXXPUSH (%s)\n", ctxpush[context-1]);
...
+                for (int i = 0; i < 32; i++) {
+                    for (int j = 0; j < nestedi[i]; j++){
+                        if (ifelse_table[i][j] != NULL){
+                            r_egg_printf(egg, "  __ifelse_%d_%d:\n", i, j);
+                            e->jmp(egg, ifelse_table[i][j], 0);
+                            R_FREE (ifelse_table[i][j]);
+                        }
+                    }
+                }
+                r_egg_printf(egg, "  __end_%d:\n\n", nfunctions);
```
Comment (radare) [inline libr/egg/egg_lang.c]: missing space

File: libr/egg/egg_lang.c [hunk 5]
```diff
@@ -559,501 +767,569 @@ eprintf ("CNTXXXPUSH (%s)\n", ctxpush[context-1]);
...
+            }
+            if (CTX>0) {
+                nbrackets++;
+            }
+            rcc_context (egg, -1);
+            if (CTX== 0) {
+                r_egg_printf(egg, "\n");
+                snprintf(str, 64, "__end_%d", nfunctions);
+                e->jmp(egg, str, 0);
+                for (int i = 0; i < 32; i++) {
```
Comment (radare) [inline libr/egg/egg_lang.c]: you cant define variables inside the for loop. this is not c++.

File: libr/egg/egg_lang.c [hunk 6]
```diff
@@ -559,501 +767,569 @@ eprintf ("CNTXXXPUSH (%s)\n", ctxpush[context-1]);
...
+                    for (int j = 0; j < nestedi[i]; j++){
+                        if (ifelse_table[i][j] != NULL){
```
Comment (radare) [inline libr/egg/egg_lang.c]: != NULL is implicit, no need to use it

File: libr/egg/egg_lang.c [hunk 7]
```diff
@@ -559,501 +767,569 @@ eprintf ("CNTXXXPUSH (%s)\n", ctxpush[context-1]);
...
+                //}
+            }
+            if (CTX>0) {
+                nbrackets++;
+            }
+            rcc_context (egg, -1);
+            if (CTX== 0) {
+                r_egg_printf(egg, "\n");
+                snprintf(str, 64, "__end_%d", nfunctions);
+                e->jmp(egg, str, 0);
```
Comment (radare) [inline libr/egg/egg_lang.c]: add missing space before (

File: libr/egg/egg_lang.c [hunk 8]
```diff
@@ -559,501 +767,569 @@ eprintf ("CNTXXXPUSH (%s)\n", ctxpush[context-1]);
...
+                r_egg_printf (egg, "  __end_%d_%d_%d:\n",
+                        nfunctions, CTX-1, nestedi[CTX-1]-1);
+                    //get_end_frame_label (egg));
+                //}
+            }
+            if (CTX>0) {
+                nbrackets++;
+            }
+            rcc_context (egg, -1);
+            if (CTX== 0) {
```
Comment (radare) [inline libr/egg/egg_lang.c]: balance spaces around ==

File: libr/egg/egg_lang.c [hunk 9]
```diff
@@ -559,501 +767,569 @@ eprintf ("CNTXXXPUSH (%s)\n", ctxpush[context-1]);
...
+                r_egg_printf(egg, "\n");
+                snprintf(str, 64, "__end_%d", nfunctions);
```
Comment (radare) [inline libr/egg/egg_lang.c]: dont hardcode constants. use sizeof(str)

File: libr/egg/egg_lang.c [hunk 10]
```diff
@@ -559,501 +767,569 @@ eprintf ("CNTXXXPUSH (%s)\n", ctxpush[context-1]);
...
+            break;
+        case '}':
+            endframe = nested[CTX];
+            if (endframe) {
+                // XXX: use endframe[context]
+                r_egg_printf (egg, "%s", endframe);
+                R_FREE (nested[CTX]);
+            //    R_FREE (endframe);
+            }
+            if (CTX>1) {
```
Comment (radare) [inline libr/egg/egg_lang.c]: spaces around >

File: libr/egg/egg_lang.c [hunk 11]
```diff
@@ -559,501 +767,569 @@ eprintf ("CNTXXXPUSH (%s)\n", ctxpush[context-1]);
...
+                if (nested_callname[CTX-1] && strstr(nested_callname[CTX-1], "if")) {
```
Comment (radare) [inline libr/egg/egg_lang.c]: space before (

File: libr/egg/egg_lang.c [hunk 12]
```diff
@@ -559,501 +767,569 @@ eprintf ("CNTXXXPUSH (%s)\n", ctxpush[context-1]);
...
+            break;
+        case '"':
+            slurp = '"';
+            break;
+        case '(':
+            slurpin = '(';
+            slurp = ')';
+            break;
+        case '{':
+            if (CTX>0) {
```
Comment (radare) [inline libr/egg/egg_lang.c]: spaces around >

File: libr/egg/egg_lang.c [hunk 13]
```diff
@@ -559,501 +767,569 @@ eprintf ("CNTXXXPUSH (%s)\n", ctxpush[context-1]);
...
+            //    r_egg_printf (egg, " %s:\n", get_frame_label (0));
+                if (nested_callname[CTX] && strstr(nested_callname[CTX], "if") &&
```
Comment (radare) [inline libr/egg/egg_lang.c]: space before (

File: libr/egg/egg_lang.c [hunk 14]
```diff
@@ -559,501 +767,569 @@ eprintf ("CNTXXXPUSH (%s)\n", ctxpush[context-1]);
...
+                    strstr(elem, "else")) {
```
Comment (radare) [inline libr/egg/egg_lang.c]: missing space, and you can put this in the line above

File: libr/egg/egg_lang.c [hunk 15]
```diff
@@ -559,501 +767,569 @@ eprintf ("CNTXXXPUSH (%s)\n", ctxpush[context-1]);
...
+                    *elem = '\x00';
+                    elem_n = 0;
+                    R_FREE (ifelse_table[CTX][nestedi[CTX]-1])
+                    snprintf (str, 64, "  __end_%d_%d_%d:",
```
Comment (radare) [inline libr/egg/egg_lang.c]: use r_str_newf()

**Comment by radare** [comment]
can you add tests in the r2r repo to verify your fixes?

**Comment by radare** [comment]
thanks for taking care of ragg :)

File: libr/egg/egg_lang.c (hunk 1 above)
Comment (ZhangZhuoSJTU) [inline libr/egg/egg_lang.c]: Ok, I will fix these indent later. Buliding problem will be fixed quickly.

**Comment by radare** [comment]
because for(int xx ) is not valid C

File: libr/egg/egg_lang.c (hunk 10 above)
Comment (ZhangZhuoSJTU) [inline libr/egg/egg_lang.c]: Actually, these missing spaces are from the original code, sad.

File: libr/egg/egg_lang.c (hunk 1 above)
Comment (ZhangZhuoSJTU) [inline libr/egg/egg_lang.c]: Sorry for my incorrect vim configure, I will fix them as fast as I can.

**Comment by ZhangZhuoSJTU** [comment]
Seems the coding style of original code are not met r2-indent or the style mentioned in the
https://github.com/radare/radare2/blob/master/CONTRIBUTING.md
I will submit the code filtered with r2-indent. Maybe r2-indent the ragg2 original code will help to review :)

File: libr/egg/egg_lang.c (hunk 15 above)
Comment (ZhangZhuoSJTU) [inline libr/egg/egg_lang.c]: I will rewrite thest "snprintf" tomorrow, sadly, for the comming final exam. Also, add some tests later.

**Pull request merged**